### PR TITLE
Overmap Game Mode Framework

### LINKED ISSUE #1595: Sol dying woefully off schedule
## Round ID: 2917

Sol gives out timers like 25 minutes til the EDF get's overun, but they always die in like 2 or 3. And on the round ID listed above they just died before we even got a warning. Ending the round instantly and out of nowhere.

## Reproduction:

Any fleet enters Sol. A decent chance of this happening occurs.

### PR DESCRIPTION
**Pull request by Karmic-Skink**
## About The Pull Request
This PR reworks a large section of SSstar_system backend to allow for modular overmap game modes to be selected independently from "normal" SS13 Gamemodes, for a more unique formula.

IE: External Gamemode + Internal GameMode
eg: Station Assault + Changeling

NOTE: This is currently a WIP, I haven't yet cleaned up a bunch of commented out code etc. Purely being PR'd now for testing.

## Why It's Good For The Game
Framework for vastly more interesting gamemodes to reduce the reliance on a single primary loop.
Fixes #1595 - Sol isn't the fail state anymore

## Changelog
:cl: Karmic Skink, Covert Corvid
refactor: Refactored entire overmap gameplay loop backend
add: Added basic versions of a couple gamemodes (not final)
/:cl:

### CODE REVIEW

**Review by Bokkiewokkie — APPROVED** [review]

**Comment by covertcorvid** [comment]
After completing an extension objective, you still get the CentCom announcement saying that a vote is pending.

File: nsv13/code/controllers/subsystem/overmap_mode.dm [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+		if(0 to 10)
+			mode.difficulty = 1
+		if(10 to 20)
+			mode.difficulty = 2
+		if(20 to 30)
+			mode.difficulty = 3
+		if(30 to 40)
+			mode.difficulty = 4
+		if(40 to INFINITY)
+			mode.difficulty = 5
```
Comment (covertcorvid) [inline nsv13/code/controllers/subsystem/overmap_mode.dm:280]: mode.difficulty = CLAMP(CEILING(players / 10), 1, 5)

**Review by covertcorvid — DISMISSED** [review]

File: code/controllers/subsystem/vote.dm [hunk 2]
```diff
@@ -149,6 +153,17 @@ SUBSYSTEM_DEF(vote)
 					var/obj/machinery/computer/communications/C = locate() in GLOB.machines
 					if(C)
 						C.post_status("shuttle")
+			if("Press On Or Return Home?") //NSV13 - Round extension vote
+				if(. == "Request Additional Objectives")
+					priority_announce("Additional Objectives") //TEMP
```
[the remark this answers is not in the document]
Comment (Karmic-Skink) [reply code/controllers/subsystem/vote.dm:158]: Pretty much yes, all those vote queues just had whatever words I could think of at that exact moment.
Likely need a pass to make them nicer.

File: nsv13/code/controllers/subsystem/overmap_mode.dm [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+	var/mode_initialised = FALSE
+
+	var/check_completion_timer = 0
+
+	var/list/mode_cache
+
+	var/list/modes
+	var/list/mode_names
+
+//some legacy vars that need to be resolved in other files
```
[the remark this answers is not in the document]
Comment (Karmic-Skink) [reply nsv13/code/controllers/subsystem/overmap_mode.dm:43]: Done.

File: nsv13/code/controllers/subsystem/overmap_mode.dm [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+
+	mode.difficulty += escalation //Our admin adjustment
+	if(mode.difficulty <= 0)
+		mode.difficulty = 1
+
+/datum/overmap_gamemode
+	var/name = null											//Name of the gamemode type
+	var/desc = null											//Description of the gamemode for ADMINS
+	var/brief = null										//Description of the gamemode for PLAYERS
+	var/config_tag = null									//Do we have a tag?
```
[the remark this answers is not in the document]
Comment (Karmic-Skink) [reply nsv13/code/controllers/subsystem/overmap_mode.dm:290]: Legacy from a previous attempt, removed.

File: nsv13/code/controllers/subsystem/overmap_mode.dm [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+		//mode_type = /datum/overmap_gamemode/patrol //Holding that as the default for now - REPLACE ME LATER
+		mode = new/datum/overmap_gamemode/patrol()
+		message_admins("Error: mode section pool empty - defaulting to PATROL")
+		log_game("Error: mode section pool empty - defaulting to PATROL")
+
+	return ..()
+
+/datum/controller/subsystem/overmap_mode/proc/setup_overmap_mode()
+	mode_initialised = TRUE
+	switch(mode.objective_reminder_setting) //Load the reminder settings
```
[the remark this answers is not in the document]
Comment (Karmic-Skink) [reply nsv13/code/controllers/subsystem/overmap_mode.dm:122]: I was thinking that those would be handled directly on the objectives but honestly I haven't really thought about how to implement it.

File: nsv13/code/controllers/subsystem/overmap_mode.dm [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+
+	text = "[text] <br><br> [mode.brief] <br><br> Objectives:"
+
+	for(var/datum/overmap_objective/O in mode.objectives)
+		text = "[text] <br> - [O.brief]"
+
+	print_command_report(text, title, TRUE)
+
+/datum/controller/subsystem/overmap_mode/proc/update_reminder(var/objective = FALSE)
+	if(objective) //Is objective? Full Reset
```
[the remark this answers is not in the document]
Comment (Karmic-Skink) [reply nsv13/code/controllers/subsystem/overmap_mode.dm:226]: Done

File: nsv13/code/controllers/subsystem/overmap_mode.dm [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+	var/combat_delay = 0									//How much time is added to the reminder timer
+	var/list/objectives = list()							//The actual gamemode objectives go here
+	var/whitelist_only = FALSE								//Can only be selected through map bound whitelists
+
+	//Reminder messages
+	var/reminder_origin = "Naval Command"
+	var/reminder_one = "This is Centcomm to all vessels assigned to explore the Delphic Expanse, please continue on your patrol route"
+	var/reminder_two = "This is Centcomm to all vessels assigned to explore the Delphic Expanse, your inactivity has been noted and will not be tolerated."
+	var/reminder_three = "This is Centcomm to all vessels assigned to explore the Delphic Expanse, we are not paying you to idle in space during your assigned patrol schedule"
+	var/reminder_four = "This is Centcomm to the explore vessel currently assigned to the Delphic Expanse, you are expected to fulfill your assigned mission"
```
[the remark this answers is not in the document]
Comment (Karmic-Skink) [reply nsv13/code/controllers/subsystem/overmap_mode.dm:308]: Fixed

File: nsv13/code/controllers/subsystem/overmap_mode.dm [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+			victory()
+			return
+		else if(O.status == STATUS_COMPLETED)
+			objective_check ++
+		else if(O.status == STATUS_FAILED)
+			objective_check ++
+			if(O.ignore_check == TRUE) //This was a gamemode objective
+				failed = TRUE
+
+	if((objective_check >= objective_length) && !failed)
```
[the remark this answers is not in the document]
Comment (Karmic-Skink) [reply nsv13/code/controllers/subsystem/overmap_mode.dm:354]: Maybe? Not sure if we want a hard failure state or just let the crew naturally explode.

File: nsv13/code/controllers/subsystem/overmap_mode.dm (hunk 1 above)
Comment (Karmic-Skink) [reply nsv13/code/controllers/subsystem/overmap_mode.dm:280]: Imagine using math.

File: nsv13/code/game/gamemodes/overmap/objectives/go_to_system.dm [hunk 9]
```diff
@@ -0,0 +1,19 @@
+/datum/overmap_objective/go_to_system
```
[the remark this answers is not in the document]
Comment (Karmic-Skink) [reply nsv13/code/game/gamemodes/overmap/objectives/go_to_system.dm:1]: I'll remove it.

File: nsv13/code/controllers/subsystem/overmap_mode.dm (hunk 8 above)
[the remark this answers is not in the document]
Comment (covertcorvid) [reply nsv13/code/controllers/subsystem/overmap_mode.dm:339]: I guess I'm mostly thinking that on the off chance they're actually good at fighting, failing becomes a way to have a non-ending round

**Comment by Bokkiewokkie** [comment]
Going to merge this now, any fixes would be easier to make later in smaller PR's or in a revert if critical.

**Pull request merged**